core/src/utils.jl

```diff
@@ -12,25 +12,73 @@ function valid_edge_types(db::DB)::Bool
     return !errors
 end
 
-"Return a directed graph, and a mapping from source and target nodes to edge fid."
-function create_graph(
-    db::DB,
-    edge_type_::String,
-)::Tuple{DiGraph, Dictionary{Tuple{Int, Int}, Int}, Dictionary{Int, Tuple{Symbol, Symbol}}}
+function create_graph(db::DB)::MetaGraph
     node_rows = execute(db, "select fid, type from Node")
-    nodes = dictionary((fid => Symbol(type) for (; fid, type) in node_rows))
-    graph = DiGraph(length(nodes))
     edge_rows = execute(db, "select fid, from_node_id, to_node_id, edge_type from Edge")
-    edge_ids = Dictionary{Tuple{Int, Int}, Int}()
-    edge_connection_types = Dictionary{Int, Tuple{Symbol, Symbol}}()
-    for (; fid, from_node_id, to_node_id, edge_type) in edge_rows
-        if edge_type == edge_type_
-            add_edge!(graph, from_node_id, to_node_id)
-            insert!(edge_ids, (from_node_id, to_node_id), fid)
-            insert!(edge_connection_types, fid, (nodes[from_node_id], nodes[to_node_id]))
+    graph = MetaGraph(
+        DiGraph();
+        label_type = NodeID,
+        vertex_data_type = NodeMetadata,
+        edge_data_type = EdgeMetadata,
+        graph_data = nothing, # In theory all remaining connectivity data could be put here
+    )
+    for row in node_rows
+        allocation_network_id =
+            hasproperty(row, :allocation_network_id) ? row.allocation_network_id : 0
+        graph[NodeID(row.fid)] =
+            NodeMetadata(Symbol(snake_case(row.type)), allocation_network_id)
+    end
+    for (; from_node_id, to_node_id, edge_type, fid) in edge_rows
+        if edge_type == "flow"
+            edge_type = EdgeType.flow
+        elseif edge_type == "control"
+            edge_type = EdgeType.control
+        else
+            error("Invalid edge type $edge_type.")
         end
+        graph[NodeID(from_node_id), NodeID(to_node_id)] =
+            EdgeMetadata(EdgeID(fid), edge_type)
     end
-    return graph, edge_ids, edge_connection_types
+    return graph
+end
+
+function inneighbor_labels_type(
```

Comment:
This and the following 2 functions might be badly named. Also, they might be allocating and thus slow.